query/plugins/org.eclipse.viatra.query.patternlanguage.emf/src/org/eclipse/viatra/query/patternlanguage/emf/validation/EMFPatternLanguageValidator.java

```diff
@@ -799,19 +799,33 @@ public void checkForWrongVariablesInXExpressions(FunctionEvaluationValue eval) {
     }
 
     private void checkForWrongVariablesInXExpressionsInternal(final XExpression expression) {
+        Pattern containingPattern = EcoreUtil2.getContainerOfType(expression, Pattern.class);
+        boolean hasSafeElementInExpressionAnnotation = PatternLanguageHelper.getFirstAnnotationByName(
+                containingPattern, SafeElementInExpressionAnnotationValidator.ANNOTATION_NAME).isPresent();
+        
         for (Variable variable : PatternLanguageHelper.getReferencedPatternVariablesOfXExpression(expression,
                 associations)) {
             IInputKey classifier = typeInferrer.getType(variable);
+            
+            String detectedError = null;
             if (classifier instanceof BottomTypeKey) {
-                error("Only simple EDataTypes are allowed in check() and eval() expressions. The variable "
-                        + variable.getName() + " has an unknown type.",
-                        expression.eContainer(), null, IssueCodes.CHECK_CONSTRAINT_SCALAR_VARIABLE_ERROR);
+                detectedError = "Only simple EDataTypes are allowed in check() and eval() expressions. The variable "
+                        + variable.getName() + " has an unknown type.";
             } else if (classifier != null && !(classifier instanceof EDataTypeInSlotsKey)
                     && !(classifier instanceof JavaTransitiveInstancesKey)) {// null-check needed, otherwise code throws
                                                                              // NPE for classifier.getName()
-                error("Only simple EDataTypes are allowed in check() and eval() expressions. The variable "
-                        + variable.getName() + " has a type of " + classifier.getPrettyPrintableName() + ".",
-                        expression.eContainer(), null, IssueCodes.CHECK_CONSTRAINT_SCALAR_VARIABLE_ERROR);
+                detectedError = "Only simple EDataTypes are allowed in check() and eval() expressions. The variable "
+                        + variable.getName() + " has a type of " + classifier.getPrettyPrintableName() + ".";
+            }
+            
+            if (null != detectedError) {
+                if (hasSafeElementInExpressionAnnotation) {
+                    info(IssueCodes.SUPPRESSED_MESSAGE_PREFIX + detectedError, 
```

Comment:
I'd argue against reporting suppressed errors as info severity notes, as they are still quite visible in the Eclipse UI. 

I'd instead do something as follows (but we might want to discuss this in more details, and possible more participants):
 1. Add an info marker to the suppression annotation (can be done using the registered AnnotationValidator instance.
 2. We could check for unnecessary suppressions as well and report them (similar to the Java @SuppressWarnings annotations in Eclipse).

Reply:
> I'd argue against reporting suppressed errors as info severity notes, as they are still quite visible in the Eclipse UI.

I think being visible is not a problem. The problem was these errors prevented compiling and loading the patterns.


> I'd instead do something as follows (but we might want to discuss this in more details, and possible more participants):
> 
>     1. Add an info marker to the suppression annotation (can be done using the registered AnnotationValidator instance.

I do not understand how this is an improvement.

>     2. We could check for unnecessary suppressions as well and report them (similar to the Java @SuppressWarnings annotations in Eclipse).

I agree that your point 2 is a nice enhancement, but should not in any case be a blocker to releasing the improvement already implemented in this PR.

Reply:
Visible error annotations with an added "[SUPPRESSED]" prefix looks very strange to me, to say the least.

1. I was not describing my intent enough here: By marking the uses of these suppression annotations as experimental, we could communicate the state of the entire feature. As an added value, if we mark these annotations explicitly, we could remove the suppression markers without any loss of information.
2. The only reason I consider it important to discuss this (I am not saying this block merging as it is) is that given our current development effort, these features tend to remain in the state they were introduced for quite some time. And I know I am late discussing the feature (but that strengthens my point).

Still, the points here are open to discussion, and I might be convinced to merge this as it is; however, the other issue in the type inferrer clearly needs to be resolved.

Reply:
> Visible error annotations with an added "[SUPPRESSED]" prefix looks very strange to me, to say the least.

Is there a way to present these that is idiomatic to Eclipse?

> By marking the uses of these suppression annotations as experimental, we could communicate the state of the entire feature

OK, that is a good call.

> 2

Do we have the capacity to implement your suggested enhancement here?
If not, I think it should be left as is - after all, we do not break any potential future implementation of this message.

Reply:
> > Visible error annotations with an added "[SUPPRESSED]" prefix looks very strange to me, to say the least.

> Is there a way to present these that is idiomatic to Eclipse?

The idiomatic way is not to present it. Think of the @SuppressWarnings annotation in Java - it completely hides the suppressed warnings.

> 2
> Do we have the capacity to implement your suggested enhancement here?
> If not, I think it should be left as is - after all, we do not break any potential future implementation of this message.

If that is the only thing not handled in this PR, I am fine with merging it, but some kind of consensus is necessary with the other points.

Reply:
> The idiomatic way is not to present it.

OK, fine.